viz/LineHolder.cpp

```diff
@@ -0,0 +1,122 @@
+#include "LineHolder.hpp"
+
+namespace vizkit3d_laser_line {
+
+LineHolder::LineHolder()
+{
+    colors = new osg::Vec4Array();
+    lineRangeVertices = new osg::Vec3Array();
+    switchNode = new osg::Switch();
+    showRange = false;
+    lineWidth = 1.0f;
+
+    setupNormalLine();
+    setupLineRange();
+}
+
+LineHolder::~LineHolder()
+{
+}
+
+void LineHolder::setupNormalLine()
+{
+    normalLineNode = new osg::Geode();
+
+    normalLineDrawable = new osg::Geometry();
+    normalLineDrawable->addPrimitiveSet(new osg::DrawArrays(osg::PrimitiveSet::LINES));
+
+    normalLineNode->addDrawable(normalLineDrawable.get());
+
+    setupStateSet(normalLineNode->getOrCreateStateSet());
+
+    switchNode->addChild(normalLineNode.get(), true);
+}
+
+void LineHolder::setupLineRange()
+{
+    lineRangeNode = new osg::Geode();
+
+    lineRangeDrawable = new osg::Geometry();
+    lineRangeDrawable->addPrimitiveSet(new osg::DrawArrays(osg::PrimitiveSet::TRIANGLE_FAN));
+
+    lineRangeNode->addDrawable(lineRangeDrawable.get());
+
+    setupStateSet(lineRangeNode->getOrCreateStateSet());
+
+    switchNode->addChild(lineRangeNode, showRange);
+
+}
+
+void LineHolder::setupStateSet(osg::StateSet *stateset)
+{
+    stateset->setAttribute(new osg::LineWidth(lineWidth), osg::StateAttribute::ON);
+    stateset->setMode(GL_LINE_SMOOTH, osg::StateAttribute::ON);
+    stateset->setMode(GL_LIGHTING, osg::StateAttribute::OFF);
+
+    // Add a blend function.
+    osg::ref_ptr<osg::BlendFunc> blend ( new osg::BlendFunc ( osg::BlendFunc::ONE, osg::BlendFunc::ONE) );
```

Comment:
This looks very weird. It's actually related to the discussion we had about references vs pointers vs const and non-const the other day.

You create a smart pointer whose job is to destroy the memory when the object it points to is unused *but* you never pass that smart pointer. Meaning either OSG has weird smart pointer semantics, or the method you call (which takes a non-const pointer) assumes it's getting ownership of the object *but* the smart pointer thinks so as well and both will delete the object eventually.

Looking at OSG code, the proper way to do this seems to be to not use a smart pointer at all

~~~
osg::BlendFunc* blend = new osg::BlendFunc ( osg::BlendFunc::ONE, osg::BlendFunc::ONE);
stateset->setAttributeAndModes ( blend, osg::StateAttribute::ON );
~~~